python/ray/data/_internal/util.py

```diff
@@ -502,37 +503,94 @@ def _split_list(arr: List[Any], num_splits: int) -> List[List[Any]]:
-    ):
-        raise ValueError(
-            "``compute`` must be specified when using a CallableClass, and must "
-            f"specify the actor compute strategy, but got: {compute}. "
-            "For example, use ``compute=ray.data.ActorPoolStrategy(size=n)``."
-        )
+    if isinstance(fn, CallableClass):
+        is_callable_class = True
+    else:
+        is_callable_class = False
+        if fn_constructor_args is not None:
+            raise ValueError(
+                "``fn_constructor_args`` can only be specified if providing a "
+                f"CallableClass instance for ``fn``, but got: {fn}"
+            )
 
-    if fn_constructor_args is not None:
-        if compute is None or (
-            compute != "actors" and not isinstance(compute, ActorPoolStrategy)
+    if compute is not None:
+        # Legacy code path to support `compute` argument.
+        logger.warning(
+            "The argument `compute` is deprecated in Ray 2.9. Please specify "
+            "argument `concurrency` instead. For more information, see "
+            "https://docs.ray.io/en/master/data/transforming-data.html#"
+            "transforming-batches-with-actors."
+        )
+        if is_callable_class and (
+            compute == "tasks" or isinstance(compute, TaskPoolStrategy)
         ):
             raise ValueError(
-                "fn_constructor_args can only be specified if using the actor "
-                f"pool compute strategy, but got: {compute}"
+                "``compute`` must be specified when using a CallableClass, and must "
+                f"specify the actor compute strategy, but got: {compute}. "
+                "For example, use ``compute=ray.data.ActorPoolStrategy(size=n)``."
             )
-        if not isinstance(fn, CallableClass):
+        elif not is_callable_class and (
+            compute == "actors" or isinstance(compute, ActorPoolStrategy)
+        ):
             raise ValueError(
-                "fn_constructor_args can only be specified if providing a "
-                f"CallableClass instance for fn, but got: {fn}"
+                "``compute`` is specified as the actor compute strategy: {compute}, "
+                "but ``fn`` is not a CallableClass: {fn}. Please remove the setting "
+                "of ``compute``."
             )
+        return compute
+    else:
+        if concurrency is not None:
+            if not is_callable_class:
+                # Currently do not support concurrency control with function,
+                # i.e., running with Ray Tasks (`TaskPoolMapOperator`).
+                raise ValueError(
+                    "``concurrency`` is specified as: {concurrency}, "
+                    "but ``fn`` is not a CallableClass: {fn}. The setting of "
+                    "``concurrency`` is only supported when ``fn`` is a CallableClass. "
+                    "This can be supported in future releases."
```

Comment:
```suggestion
                    "``concurrency`` is set, "
                    "but ``fn`` is not a CallableClass: {fn}. Non-default values of ``concurrency`` are currently only supported when ``fn`` is a CallableClass."
```

(I think it's fine not to mention and this is more concise)